src/lib.rs

```diff
@@ -32,44 +32,69 @@ lalrpop_mod!(#[allow(clippy::all)] pub parser);
 /// Compile a complete system policy
 /// The list of input files list should contain filenames of files containing policy to be
 /// compiled.
+/// The list of system names are the names of the systems to build.
 /// Returns a Result containing either a string of CIL policy which is the compiled result or a
 /// list of errors.
-/// In order to convert the compiled CIL policy into a usable policy, you must use secilc
-pub fn compile_system_policy(input_files: Vec<&str>) -> Result<String, error::CascadeErrors> {
-    let mut policies: Vec<PolicyFile> = Vec::new();
-    let mut errors = CascadeErrors::new();
+/// In order to convert the compiled CIL policy into a usable policy, you must use secilc.
+pub fn compile_system_policies(
+    input_files: Vec<&str>,
+    system_names: Vec<String>,
+) -> Result<HashMap<String, String>, error::CascadeErrors> {
+    let policies = get_policies(input_files)?;
+    compile_system_policies_internal(policies, system_names, false)
+}
 
-    for f in input_files {
-        let policy_str = match std::fs::read_to_string(&f) {
-            Ok(s) => s,
-            Err(e) => {
-                errors.add_error(e);
-                continue;
-            }
-        };
-        let p = match parse_policy(&policy_str) {
-            Ok(p) => p,
-            Err(evec) => {
-                for e in evec {
-                    // TODO: avoid String duplication
-                    errors.add_error(error::ParseError::new(e, f.into(), policy_str.clone()));
-                }
-                continue;
+/// Compile all of the system policies
+/// The list of input files list should contain filenames of files containing policy to be
+/// compiled.
+/// Returns a Result containing either a string of CIL policy which is the compiled result or a
+/// list of errors.
+/// In order to convert the compiled CIL policy into a usable policy, you must use secilc.
+pub fn compile_system_policies_all(
+    input_files: Vec<&str>,
+) -> Result<HashMap<String, String>, error::CascadeErrors> {
+    let mut system_names = Vec::new();
+    let policies = get_policies(input_files)?;
+    for p in &policies {
+        for e in &p.policy.exprs {
+            if let Expression::Decl(Declaration::System(s)) = e {
+                system_names.push(s.name.to_string());
             }
-        };
-
-        policies.push(PolicyFile::new(*p, SimpleFile::new(f.into(), policy_str)));
+        }
     }
-    // Stops if something went wrong for this major step.
-    errors = errors.into_result_self()?;
+    compile_system_policies_internal(policies, system_names, false)
+}
+
+/// Compile all systems into a single policy
+/// The list of input files list should contain filenames of files containing policy to be
+/// compiled.
+/// The system name is the name of the output file (out.cil by default).
```

Comment:
The default is being supplied by casc, not the library, right?  So calling it out here at the casc/library boundary doesn't make sense.  It looks like the code just clones whatever system_name is passed in.  Also, the library doesn't generate a file, it returns the string that is the contents of the file and casc actually makes the file.

I think the code makes sense, and the comment is what need changing.

Reply:
I'm looking more at the usages in the tests and in main.rs.  I think there's some weirdness between "output file name" and "system name".

In `compile_system_policies` and `compile_system_policies_all` the system names are the portion of the filename without .cil (eg "out").  In `compile_combined`, the system name is the full file.  Then casc accounts for that by appending ".cil" in the first two cases, but not in compile_combined.

I think the library interface needs to be consistent here.  System names are always just the descriptive name, without ".cil", and the output file name will have ".cil" appended to it.  That means that in casc, the call to compile_combined will need to get the portion of the output filename before .cil, and then append .cil afterwards.

Reply:
Actually, why do we care about this at all?  casc sets the output file name.  Does compile_combined even need to know it?  All compile combined needs is some internal name to use for the system, but that shouldn't end up affecting what the user gets, since we're not actually returning the system name, right?  Can we not just drop the system_name argument here entirely and just use some internal value?